File: zone-manager/zm-client/src/app/model/Owner.ts
import {Urban_Zone} from "./Organization/Urban_Zone";

export enum OwnerRole{
  ADMINISTRATOR="ADMINISTRATOR",
  SERVICE_PROVIDER="SERVICE_PROVIDER",
  BUSINESS_OWNER="BUSINESS_OWNER",
  OWNER="OWNER",
  NONE="NONE",
}

export class Owner {
  constructor(
    public id: string = "",
    public username: string = "",
    public email: string = "",
    public phone: string = "",
    public role: OwnerRole = OwnerRole.NONE,
    public zones: Urban_Zone[] = []
  ) {}
}

export class OwnerDto extends Owner{
  public password: string = "";
  constructor(owner: Owner, password: string = '') {
    super(
      owner.id,
      owner.username,
      owner.email,
      owner.phone,
      owner.role,
      owner.zones
    )
    this.password = password
  }
}

File: zone-manager/zm-server/src/main/kotlin/com/uims/zm/zonemanager/owner/entity/Owner.kt
package com.uims.zm.zonemanager.owner.entity

import jakarta.persistence.*
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import java.util.*

@Entity
@Table(name = "owner")
 class Owner : UserDetails {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "id", nullable = false)
    open var id: UUID? = null

    @Column(name = "username", nullable = false)
    private var username: String = "Admim - " + UUID.randomUUID()

    @Column(name = "password", nullable = false)
    private var password: String = ""

    @Column(name = "email", nullable = false)
    open var email: String = ""

    @Column(name = "phone", nullable = false)
    open var phone: String = ""

    @Enumerated(EnumType.STRING)
    @Column(name = "role", nullable = false)
    open var role: OwnerRole = OwnerRole.OWNER

    override fun getAuthorities(): MutableCollection<out GrantedAuthority> {
        return mutableListOf(SimpleGrantedAuthority(this.role.toString()));
    }

    override fun getPassword(): String {
        return this.password
    }

    override fun getUsername(): String {
        return this.username
    }

    override fun isAccountNonExpired(): Boolean {
        return true;
    }

    override fun isAccountNonLocked(): Boolean {
        return true;
    }

    override fun isCredentialsNonExpired(): Boolean {
        return true;
    }

    override fun isEnabled(): Boolean {
        return true;
    }

    fun setEncodedPassword(encodedPassword: String){
        this.password = encodedPassword;
    }
}